### PR TITLE
[5.9] Convert Arrayable Eloquent attributes to Array

### PR DESCRIPTION
**Pull request by mathieutu**
Hi, 
If you cast a json field to a collection, or have mutators/getters which return Arrayable object in a Eloquent model, the `toArray()` method will not convert them to array recursively as it is supposed to do (it has this behavior in collections, and for model's relations).

This PR fix that.

I'll add some tests later, if you agree with this PR.

I'm wondering about the targeted version. I would initially target 5.8, as it's a fix, but It could be considerate as a BC, because it changes the behavior of this toArray() method concerning nested Arrayable objects.

So maybe 6.0? I let you take the decision 😉.

Thanks, 
Matt.

### CODE REVIEW

**Comment by mathieutu** [comment]
I found this issue which is fixed by this PR: https://github.com/laravel/framework/issues/19082 (ping @themsaid)

**Comment by taylorotwell** [comment]
Hmm, it is a little unfortunate that we would now have to spin through every attribute for every model when converting a collection to an array.

**Comment by mathieutu** [comment]
Ok, I understand and know how to avoid that: as only casted attributes can be Arrayable, we can integrate this process in the casting part. Will push an update soon.

**Comment by mathieutu** [comment]
@taylorotwell This should be much better, no more loop! 
This was actually already handled for mutatedAttributes, so only the casts ones need it.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -195,6 +195,11 @@ protected function addCastAttributesToArray(array $attributes, array $mutatedAtt
             if ($attributes[$key] && $this->isCustomDateTimeCast($value)) {
                 $attributes[$key] = $attributes[$key]->format(explode(':', $value, 2)[1]);
             }
+
+            if ($attributes[$key] instanceof Arrayable) {
+                $attributes[$key] = $attributes[$key]->toArray();
```
Comment (mathieutu) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]:
```suggestion
                $attributes[$key] = $attributes[$key]->toArray();
```

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 2]
```diff
@@ -195,6 +195,11 @@ protected function addCastAttributesToArray(array $attributes, array $mutatedAtt
             if ($attributes[$key] && $this->isCustomDateTimeCast($value)) {
                 $attributes[$key] = $attributes[$key]->format(explode(':', $value, 2)[1]);
             }
+
+            if ($attributes[$key] instanceof Arrayable) {
+                $attributes[$key] = $attributes[$key]->toArray(); 
+            }
+
```
Comment (mathieutu) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]:
```suggestion
```

**Comment by taylorotwell** [comment]
Overall that looks good. I think it may be better to target it to master though. However, one question I do have is should this conversion even happen? Do we cast all objects to array using (array)?

**Comment by mathieutu** [comment]
@taylorotwell We do not use `(array)` cast, because if we do `(array) 'foo'` will become `['foo']`. However IMO `toArray` methods call with nested `Arrayable` items should call the method on them. Example with collection, paginator, etc.

On the model it's already the case for the mutated attributes. if getXXAttribute return an `Arrayable` object, it will be cast to array. 

For casts, the other objects are already converted to native type. As an example, Date is exported as a string instead of keeping the Carbon type.

Natively the only cast with Arrayable contract is the collection. For me it's a bit weird to call `toArray` and have a nested collection in fine. 

I'll target master and add some tests if you're ok with that.

Thanks.

**Comment by taylorotwell** [comment]
Yes, please add tests.

**Comment by taylorotwell** [comment]
Closing pending tests.

**Pull request closed without merging**